pkg/slo-controller/noderesource/plugins/resourceamplification/plugin.go

```diff
@@ -0,0 +1,112 @@
+	"github.com/koordinator-sh/koordinator/pkg/slo-controller/noderesource/framework"
+)
+
+const PluginName = "ResourceAmplification"
+
+// Plugin calculates and updates final node resource amplification ratios automatically
+// based on user config (not implemented) and node cpu normalization ratio.
+type Plugin struct{}
+
+func (p *Plugin) Name() string {
+	return PluginName
+}
+
+func (p *Plugin) NeedSyncMeta(_ *configuration.ColocationStrategy, oldNode, newNode *corev1.Node) (bool, string) {
+	oldRatioStr := oldNode.Annotations[extension.AnnotationNodeResourceAmplificationRatio]
+	newRatioStr := newNode.Annotations[extension.AnnotationNodeResourceAmplificationRatio]
+
+	if oldRatioStr == "" && newRatioStr == "" {
+		return false, "ratio remains empty"
+	}
+	if oldRatioStr == "" {
+		return true, "old ratio is empty"
+	}
+	if newRatioStr == "" {
+		return true, "new ratio is empty"
+	}
+	if oldRatioStr == newRatioStr {
+		return false, "ratio remains unchanged"
+	}
+	return true, "ratio changed"
+}
+
+func (p *Plugin) Prepare(_ *configuration.ColocationStrategy, node *corev1.Node, nr *framework.NodeResource) error {
+	ratioStr, ok := nr.Annotations[extension.AnnotationNodeResourceAmplificationRatio]
+	if !ok {
+		klog.V(6).Infof("prepare node resource amplification ratio to unset, node %s", node.Name)
+		delete(node.Annotations, extension.AnnotationNodeResourceAmplificationRatio)
+		return nil
+	}
+
+	if node.Annotations == nil {
+		node.Annotations = map[string]string{}
+	}
+	node.Annotations[extension.AnnotationNodeResourceAmplificationRatio] = ratioStr
+	klog.V(6).Infof("prepare node resource amplification ratio to set, node %s, ratio %s", node.Name, ratioStr)
+
+	return nil
+}
+
+func (p *Plugin) Reset(node *corev1.Node, message string) []framework.ResourceItem {
+	// Currently we have no user configurations so there's no need to reset.
+	return nil
+}
+
+func (p *Plugin) Calculate(_ *configuration.ColocationStrategy, node *corev1.Node, _ *corev1.PodList, _ *framework.ResourceMetrics) ([]framework.ResourceItem, error) {
+	// Set cpu amplification ratio according to cpu normalization ratio.
+	// In the future, we should read user's amplification config and multiply its cpu amplification ratio with cpu normalization ratio.
+	normRatio, err := extension.GetCPUNormalizationRatio(node)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get cpu normalization ratio: %w", err)
+	}
+
+	if normRatio <= 1 {
+		return []framework.ResourceItem{
+			{
+				Name: PluginName,
+			},
+		}, nil
+	}
+
+	ampRatios := map[corev1.ResourceName]extension.Ratio{
+		corev1.ResourceCPU: extension.Ratio(normRatio),
+	}
+	ratioBytes, _ := json.Marshal(ampRatios)
+	ratioStr := string(ratioBytes)
+	klog.V(6).Infof("calculate resource amplification ratio %s for node %s", ratioStr, node.Name)
+	return []framework.ResourceItem{
+		{
+			Name: PluginName,
+			Annotations: map[string]string{
+				extension.AnnotationNodeResourceAmplificationRatio: ratioStr,
```

Comment:
If AnnotationNodeResourceAmplificationRatio already exists and the value is {"memory": "1.2"}, then the memory will be erased here.

Reply:
In current design, the `AnnotationNodeResourceAmplificationRatio` is fully managed by this plugin. In the future, we shall introduce a new annotation or other configuration methods for users and this plugin would take these into consideration when calculating the final `AnnotationNodeResourceAmplificationRatio`.

Reply:
I've updated the comment to make this intention clearer.

Reply:
ok.